### PR TITLE
Save last working directory

### PR DESCRIPTION
**Pull request by ppwadhwa**
# Description

When opening or saving files, the last directory you saved to or opened from will be the directory you start from the next time you open an `open` or `save` file dialog.  

The most recent history seems to already be there in the recent places list on mac without using setHistory:
Screen Shot 2021-03-24 at 6 55 24 PM

Work from issue:
https://github.com/napari/product-heuristics-2020/issues/32

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by goanpeca** [comment]
Hi @ppwadhwa thanks for working on this.

>The most recent history seems to already be there in the recent places list on mac without using setHistory:

Indeed, but I think that this relates to global OS, that means that usiong other programs will change those values. 

We might want to store the last X number of files, and make sure the Napari dialogs show those values after closing and opening (and using the OS dialogs on other programs)

Now the issue I see is that we are using the [static dialog Helpers everywhere](https://doc.qt.io/qt-5/qfiledialog.html#static-public-members) everywhere, so we would need to use QFileDialog directory, in order to be able to use `setHistory`, since the static members produce a modal, and it seems set history is not part of the args.

Some questions for @napari/core-devs 
- Do we want to set this history separately for Open and Save, or do we want the same?
- To be able to store the history, we will need to replace the static member calls with normal QFileDialog, creation, or we could create our own helpers in a separate module that accepted the history as an argument.

Thoughts?

**Comment by sofroniewn** [comment]
> Do we want to set this history separately for Open and Save, or do we want the same?

I will defer to recommendations from @jni and @liaprins-czi on what is standard and best practice, but I think if this PR is just focused on directories then the same for open and save is ok with me.

I just want to check thought that you see this as separate from #2447, which seeks to save a list of the most recently open files? There I think it is important it is opened files only.

> To be able to store the history, we will need to replace the static member calls with normal QFileDialog, creation, or we could create our own helpers in a separate module that accepted the history as an argument.

I'm pretty open on what is best practices here. I may have added some of the helpers and not quite known what i was doing. @tlambert03 has more expertise than me here so might be able to provide a better recommendation

**Comment by liaprins-czi** [comment]
> Do we want to set this history separately for Open and Save, or do we want the same?

Great question. The default location that opens for either action should be different:
- For saving, the default directory should be where the image currently lives / was opened from.
- For opening, the default directory should be the directory most recently opened from.
- However, for the list of recent places, it can and should be a combination: the most recently opened-from + most recently saved-to folders.

**Comment by goanpeca** [comment]
@ppwadhwa and @sofroniewn we could have something like this, to include history, but retain the convenience of using the status members of QFileDialog

```python
# napari/_qt/utils.py
def getExistingDirectory(parent, caption=None, directory=None, history=None):
    """
    A helper method that basically mimics QFileDialog.getExistingDirectory, but adds a history parameter.
    """
    dlg = QFileDialog(parent)
    # this is mock code.. 
    dlg.setOptions(DirsOnlys)
    dlg.setCaption(caption)
    if history:
        dlg.setHistory(history)
    return dlg.exec_()

# napari/utils/io.py?  napari/utils/misc.py? or napari/utils/history.py
def update_open_history(filenames):
    """A helper for history handling."""
    folders = SETTINGS.application.open_history
    folders = [f for f in folders if os.path.isdir(f)]
    return folders


def get_open_history():
    """A helper for history handling."""
    folders = SETTINGS.application.open_history
    folders = [f for f in folders if os.path.isdir(f)]
    return folders[0]

# napari/_qt/qt_viewer.py
# Change QFileDialog.getExistingDirectory to getExistingDirectory

from .utils import getExistingDirectory

    # ....
    getExistingDirectory(
            parent=self,
            caption=trans._('Select folder...'),
            directory=SETTINGS.application.last_visited_dir,  # home dir by default
            history=get_open_history()
        )
```

Thoughts?

---

Thanks for the comments @liaprins-czi , I do not follow this one

>For saving, the default directory should be where the image currently lives / was opened from.

Napari does not have the concept of a current file that it may link to in disk (like a word document or similar)

**Comment by sofroniewn** [comment]
> @ppwadhwa and @sofroniewn we could have something like this, to include history, but retain the convenience of using the status members of QFileDialog

I'm fine with this approach, but maybe @tlambert03 can weigh in too

**Comment by tlambert03** [comment]
> we will need to replace the static member calls with normal QFileDialog, creation, or we could create our own helpers in a separate module that accepted the history as an argument.

I have no problems at all with replacing many or all of our static calls with a normal QFileDialog.  The main thing I wanted to avoid earlier on was a proliferation of single-use subclasses (`SaveFileDialog`, `OpenImageDialog`, `GetPluginDialog`, etc...) where there was no real need to actually create a custom subclass.  But there was nothing we were specifically aiming for by using the static members.  Any changes there are fine with me (don't really care whether we create a convenience function mimicking the static calls either).

**Comment by liaprins-czi** [comment]
@goanpeca Re
> Napari does not have the concept of a current file that it may link to in disk (like a word document or similar)

In that case, can we open to the most recently saved-to location by default?

**Comment by ppwadhwa** [comment]
I've made some changes to make sure that there is a history of save directories and a history of opened directories.  I am creating a dialog object in each case instead of using helper functions.  To me that seemed simpler, but I can change it if desired.  Also, the history is being saved as I expect, but I'm still not convinced that the `setHistory` function is doing its job on my mac.  I will play some more with this tomorrow to see if I can convince myself one way or another.

**Comment by goanpeca** [comment]
Thanks for the updates @ppwadhwa !

--- 

@Czaki could you check if this PR is updating the history for you?

Thanks!

**Comment by Czaki** [comment]
I will check this tomorrow, but after fast review I would like to ask why its uses a global state for it? This even does not use `Settings` object but some custom functions. 

There where to many cases where I need to fight with global state. Increasing number of places where global state is stored may introduce a really big headache in the future.

**Comment by goanpeca** [comment]
> but after fast review I would like to ask why its uses a global state for it? This even does not use Settings object but some custom functions.

Not sure I understand this.

>There where to many cases where I need to fight with global state. Increasing number of places where global state is stored may introduce a really big headache in the future.

Are you talking about `SETTINGS` ? yes we will move away from a Global object, but that will happen on another PR :) 

>I will check this tomorrow,

Thanks!

**Comment by Czaki** [comment]
History is obtained using:
```
hist = get_save_history()
```

which means that it is stored in some global object and does not shortly connected to SETTINGS object.

**Comment by ppwadhwa** [comment]
@Czaki Thank you so much for taking the time to look this over!  I may be misunderstanding your concern, and if so I apologize, but I wanted to point out that `get_save_history` is a helper function that pulls the history saved in `SETTINGS`.  Also, if you use `update_save_history`, this function will update `SETTINGS.application.save_history`.

**Comment by Czaki** [comment]
I see that it uses SETTINGS internally, but, as you mentioned there is a plan to go out of global object so why settings are not passed as an argument of these functions or these functions are not implemented as a method of SETTINGS object?

**Comment by goanpeca** [comment]
>I see that it uses SETTINGS internally, but, as you mentioned there is a plan to go out of global object so why settings are not passed as an argument of these functions or these functions are not implemented as a method of SETTINGS object?

When we do that we will basically search `SETTINGS` and fix as needed so having less `SETTINGS` around the codebase right now will make that process easier. That is why we have it only on the functions and not where they are used, since we still need to define how we are going to be propagating a settings object in the main window and viewers. 

---

On another note, did you get a chance to test if the PR works for you as expected?

Thanks for the feedback :)

File: napari/utils/history.py
Comment (sofroniewn) [inline napari/utils/history.py]: Should this stay or go? If not needed, maybe you can drop before we merge

File: napari/utils/history.py
Comment (sofroniewn) [inline napari/utils/history.py]: Should this stay or go? If not needed, maybe you can drop before we merge

File: napari/utils/history.py
Comment (sofroniewn) [inline napari/utils/history.py]: Can you add a `Parameters` section to this doc string with `filename` in it

File: napari/utils/history.py
Comment (sofroniewn) [inline napari/utils/history.py:24]: Can you add a `Parameters` section to this doc string with `filename` in it

**Review by sofroniewn — APPROVED** [review]
Looks great! Worked as expected for me. Left some very minor review comments, but otherwise good to go!!

**Review by goanpeca — APPROVED** [review]
Great work @ppwadhwa!

I Agree with @sofroniewn comments and suggestions :-)

File: napari/_qt/qt_viewer.py
Comment (ppwadhwa) [inline napari/_qt/qt_viewer.py:488]: I recently moved this save history function to here to make sure it updated the history when a file was successfully saved.

**Comment by ppwadhwa** [comment]
@sofroniewn thanks for catching the unnecessary comments and the need for parameters in the doctsrings!  I added them.  I also made one small change to the place where the `update_save_history` was being called.  I tested the behavior of this PR one more time and it was being finicky on some of the save history, so I moved that one line.  I just made a note by it so you can more easily find where I made this recent change.  

also, I did a rebase with master and an `os` import slipped in that is not being used in `qt_main_window.py`.  That is why one of the tests are failing.  I didn't add it (it came with the rebase), but I can remove it if you want.  I didn't want to remove it without permission since I may be missing the reason its there in the first place.

**Comment by sofroniewn** [comment]
> also, I did a rebase with master and an os import slipped in that is not being used in qt_main_window.py. That is why one of the tests are failing. I didn't add it (it came with the rebase), but I can remove it if you want. I didn't want to remove it without permission since I may be missing the reason its there in the first place.

Looks like we used to need it, but you're cutting a line with `os.path.dirname`. I think fine to drop if you can't see it being used elsewhere and all the tests pass/ functionality works as expected. @tlambert03 might remember something else that I'm missing, but dropping seems fine to me!!

**Comment by ppwadhwa** [comment]
oh!  yes, I see.  ok, I will remove it.

**Comment by sofroniewn** [comment]
Ok nice, I think this is good to go in, will merge tomorrow. Thanks!!

**Pull request merged**
